lib/dataset/sized_dict.cpp

```diff
@@ -348,56 +318,237 @@ SizedDict<Key, Value>::merge_from(const SizedDict &other) const {
-  core::Dict<Key, Value> out;
+AlignedDict<Key, Value>::AlignedDict(
+    const Sizes &sizes,
+    std::initializer_list<std::pair<const Key, Value>> items,
+    const bool readonly)
+    : AlignedDict(sizes, raw_holder_type(items), readonly) {}
+
+template <class Key, class Value>
+AlignedDict<Key, Value>::AlignedDict(Sizes sizes, raw_holder_type items,
+                                     const bool readonly) {
+  this->m_sizes = std::move(sizes);
+  for (auto &&[key, value] : items)
+    set(key, std::move(value));
+  // `set` requires Dict to be writable, set readonly flag at the end.
+  this->m_readonly =
+      readonly; // NOLINT(cppcoreguidelines-prefer-member-initializer)
+}
+
+template <class Key, class Value>
+AlignedDict<Key, Value>::AlignedDict(Sizes sizes, holder_type items,
+                                     const bool readonly) {
+  this->m_sizes = std::move(sizes);
+  for (auto &&[key, value] : items)
+    set(key, std::move(value.value), value.aligned);
+  // `set` requires Dict to be writable, set readonly flag at the end.
+  this->m_readonly =
+      readonly; // NOLINT(cppcoreguidelines-prefer-member-initializer)
+}
+
+template <class Key, class Value>
+AlignedDict<Key, Value>::AlignedDict(
+    const SizedDict<Key, AlignedValue<Value>, AlignedDict<Key, Value>> &other)
+    : SizedDict<Key, AlignedValue<Value>, AlignedDict<Key, Value>>{other} {}
+
+template <class Key, class Value>
+AlignedDict<Key, Value>::AlignedDict(
+    SizedDict<Key, AlignedValue<Value>, AlignedDict<Key, Value>>
+        &&other) noexcept
+    : SizedDict<Key, AlignedValue<Value>, AlignedDict<Key, Value>>{
+          std::move(other)} {}
+
+template <class Key, class Value>
+AlignedDict<Key, Value>::AlignedDict(const AlignedDict &other)
+    : AlignedDict(other.m_sizes, other.m_items, false) {}
+
+template <class Key, class Value>
+AlignedDict<Key, Value>::AlignedDict(AlignedDict &&other) noexcept = default;
+
+template <class Key, class Value>
+AlignedDict<Key, Value> &
+AlignedDict<Key, Value>::operator=(const AlignedDict &other) = default;
+
+template <class Key, class Value>
+AlignedDict<Key, Value> &
+AlignedDict<Key, Value>::operator=(AlignedDict &&other) noexcept = default;
+
+/// Const reference to the coordinate for given dimension.
+template <class Key, class Value>
+const Value &AlignedDict<Key, Value>::operator[](const Key &key) const {
+  return at(key);
+}
+
+/// Const reference to the coordinate for given dimension.
+template <class Key, class Value>
+const Value &AlignedDict<Key, Value>::at(const Key &key) const {
+  return raw_at(key).value;
+}
+
+/// The coordinate for given dimension.
+template <class Key, class Value>
+Value AlignedDict<Key, Value>::operator[](const Key &key) {
+  return std::as_const(*this).at(key);
+}
+
+template <class Key, class Value>
+const AlignedValue<Value> &
+AlignedDict<Key, Value>::raw_at(const Key &key) const {
+  scipp::expect::contains(*this, key);
+  return this->m_items.at(key);
```

Comment:
Not sure it would matter in practice, but why are we not using `find` and then return, to avoid doing two lookups?

Reply:
Lazyness? `contains` takes care of formatting the error message. But I agree, this should be done with a single iteration.